chainer/function.py

```diff
@@ -622,11 +627,33 @@ def __enter__(self):
             raise KeyError('hook %s already exists' % self.name)
 
         function_hooks[self.name] = self
+        self.added()
         return self
 
     def __exit__(self, *_):
+        chainer.get_function_hooks()[self.name].deleted()
         del chainer.get_function_hooks()[self.name]
 
+    def added(self, function=None):
+        """Callback function invoked when the hook is added.
+
+        Args:
+            function(~chainer.Function): Function object to which
+                the function hook is registered. None if the hook
+                is registered via ``with`` statement.
+        """
+        pass
+
+    def deleted(self, function=None):
```

Comment:
I needed this additional hook to successfully clean cupy memory hook registered via both `add_hook` and `__enter__` (with statement).